From 35f3372fa2b1deb6af7bb3e34e7ebcbeff6a874f Mon Sep 17 00:00:00 2001
Message-Id: <35f3372fa2b1deb6af7bb3e34e7ebcbeff6a874f.1413837856.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Gajanan Bhat <gajanan.bhat@intel.com>
Date: Wed, 15 May 2013 11:22:05 +0530
Subject: [PATCH 027/312] FOR_UPSTREAM [VPG]: drm/i915: Falling back to bit
 banging if GMBUS returns NAK

On multiple VLV boards GMBUS failure was observed. Only when GMBUS timeout
occurs the fallback to i2c is triggered. Need to enable i2c fallback in
any GMBUS failures. Bit banging was used only in case of GMBUS timeout.
This patch enables use of i2c in case of NAK and timeout of GMBUS.

Issue: APDEV-1107
OTC-Tracker: VIZ-3458
Change-Id: I9d8c8c1de927c621f8991450643532447260eac6
Signed-off-by: Gajanan Bhat <gajanan.bhat@intel.com>
Reviewed-on: http://android.intel.com:8080/107667
Reviewed-by: cactus <cactus@intel.com>
Reviewed-by: Purushothaman, Vijay A <vijay.a.purushothaman@intel.com>
Reviewed-by: Mohan Marimuthu, Yogesh <yogesh.mohan.marimuthu@intel.com>
Reviewed-by: Stimson, Dale B <dale.b.stimson@intel.com>
Tested-by: Stimson, Dale B <dale.b.stimson@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
Signed-off-by: A.Sunil Kamath <sunil.kamath@intel.com>
(cherry picked from commit 1baa38519038d3f21b554113656ea6f1cae0aded)

Signed-off-by: Jon Bloomfield <jon.bloomfield@intel.com>
---
 drivers/gpu/drm/i915/intel_i2c.c |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/intel_i2c.c b/drivers/gpu/drm/i915/intel_i2c.c
index d33b61d..dd9f2dc 100644
--- a/drivers/gpu/drm/i915/intel_i2c.c
+++ b/drivers/gpu/drm/i915/intel_i2c.c
@@ -533,11 +533,15 @@ clear_err:
 			 adapter->name, msgs[i].addr,
 			 (msgs[i].flags & I2C_M_RD) ? 'r' : 'w', msgs[i].len);
 
-	goto out;
+	goto fallback;
 
 timeout:
 	DRM_INFO("GMBUS [%s] timed out, falling back to bit banging on pin %d\n",
 		 bus->adapter.name, bus->reg0 & 0xff);
+	DRM_INFO("GMBUS [%s] timed out,\n", bus->adapter.name);
+
+fallback:
+	DRM_INFO("Falling back to bit banging on pin %d\n", bus->reg0 & 0xff);
 	I915_WRITE(GMBUS0 + reg_offset, 0);
 
 	/* Hardware may not support GMBUS over these pins? Try GPIO bitbanging instead. */
-- 
1.7.9.5

